// @conversationlearner/sdk/lib/Memory/ClientMemoryManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const CLDebug_1 = require("../CLDebug");
const models_1 = require("@conversationlearner/models");
class ClientMemoryManager {
    constructor(prevMemories, curMemories, allEntities, sessionInfo) {
        this.allEntities = [];
        this.allEntities = allEntities;
        this.sessionInfo = sessionInfo;
        this.prevMemories = prevMemories;
        this.curMemories = curMemories;
    }
    FindEntity(entityName) {
        let match = this.allEntities.find(e => e.entityName == entityName);
        return match;
    }
    RememberEntity(entityName, entityValue) {
        let entity = this.FindEntity(entityName);
        if (!entity) {
            CLDebug_1.CLDebug.Error(`Can't find Entity named: ${entityName}`);
            return;
        }
        if (entity.entityType != models_1.EntityType.LOCAL) {
            CLDebug_1.CLDebug.Error(`Not allowed to set values of pre-built Entities: ${entityName}`);
            return;
        }
        if (typeof entityValue == 'object') {
            entityValue = JSON.stringify(entityValue);
        }
        else if (typeof entityValue == 'number') {
            entityValue = entityValue.toString();
        }
        this.curMemories.Remember(entity.entityName, entity.entityId, entityValue, entity.isMultivalue);
    }
    RememberEntities(entityName, entityValues) {
        let entity = this.FindEntity(entityName);
        if (!entity) {
            CLDebug_1.CLDebug.Error(`Can't find Entity named: ${entityName}`);
            return;
        }
        if (entity.entityType != models_1.EntityType.LOCAL) {
            CLDebug_1.CLDebug.Error(`Not allowed to set values of pre-built Entities: ${entityName}`);
            return;
        }
        if (!entity.isMultivalue) {
            CLDebug_1.CLDebug.Error(`RememberEntitiesAsync called on entity (${entityName}) that isn't Multi-Value.  Only the last value will be remembered`);
        }
        this.curMemories.RememberMany(entity.entityName, entity.entityId, entityValues, entity.isMultivalue);
    }
    ForgetEntity(entityName, value = null) {
        let entity = this.FindEntity(entityName);
        if (!entity) {
            CLDebug_1.CLDebug.Error(`Can't find Entity named: ${entityName}`);
            return;
        }
        // If no value given, wipe all entites from buckets
        this.curMemories.Forget(entity.entityName, value, entity.isMultivalue);
    }
    /** Clear all entity values apart from any included in the list of saveEntityNames
     * Useful in the "onSessionEndCallback" to preserve a subset of entities for the next session
     */
    ForgetAllEntitiesAsync(saveEntityNames) {
        for (let entity of this.allEntities) {
            if (saveEntityNames.indexOf(entity.entityName) < 0) {
                this.curMemories.Forget(entity.entityName, null, entity.isMultivalue);
            }
        }
    }
    CopyEntity(entityNameFrom, entityNameTo) {
        let entityFrom = this.FindEntity(entityNameFrom);
        let entityTo = this.FindEntity(entityNameTo);
        if (!entityFrom) {
            CLDebug_1.CLDebug.Error(`Can't find Entity named: ${entityNameFrom}`);
            return;
        }
        if (!entityTo) {
            CLDebug_1.CLDebug.Error(`Can't find Entity named: ${entityNameTo}`);
            return;
        }
        if (entityFrom.isMultivalue != entityTo.isMultivalue) {
            CLDebug_1.CLDebug.Error(`Can't copy between Bucket and Non-Bucket Entities`);
            return;
        }
        // Clear "To" entity
        this.curMemories.Forget(entityNameTo);
        // Get value of "From" entity
        let values = this.curMemories.ValueAsList(entityNameFrom);
        // Copy values from "From"
        for (let value of values) {
            this.RememberEntity(entityNameTo, value);
        }
    }
    EntityValue(entityName) {
        return this.curMemories.ValueAsString(entityName);
    }
    PrevEntityValue(entityName) {
        return this.prevMemories.ValueAsString(entityName);
    }
    EntityValueAsPrebuilt(entityName) {
        return this.curMemories.ValueAsPrebuilt(entityName);
    }
    PrevEntityValueAsPrebuilt(entityName) {
        return this.prevMemories.ValueAsPrebuilt(entityName);
    }
    EntityValueAsList(entityName) {
        return this.curMemories.ValueAsList(entityName);
    }
    PrevEntityValueAsList(entityName) {
        return this.prevMemories.ValueAsList(entityName);
    }
    EntityValueAsNumber(entityName) {
        return this.curMemories.ValueAsNumber(entityName);
    }
    PrevValueAsNumber(entityName) {
        return this.prevMemories.ValueAsNumber(entityName);
    }
    EntityValueAsBoolean(entityName) {
        return this.curMemories.ValueAsBoolean(entityName);
    }
    PrevValueAsBoolean(entityName) {
        return this.prevMemories.ValueAsBoolean(entityName);
    }
    EntityValueAsObject(entityName) {
        return this.curMemories.ValueAsObject(entityName);
    }
    PrevEntityValueAsObject(entityName) {
        return this.prevMemories.ValueAsObject(entityName);
    }
    GetFilledEntitiesAsync() {
        return this.curMemories.FilledEntities();
    }
    SessionInfo() {
        return this.sessionInfo;
    }
}
exports.ClientMemoryManager = ClientMemoryManager;
