// @conversationlearner/sdk/lib/AzureFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
const Request = require("request");
const CLDebug_1 = require("./CLDebug");
class AzureFunctions {
    static Call(azureFunctionsUrl, azureFunctionsKey, funcName, args) {
        let apiPath = 'app';
        if (azureFunctionsKey) {
            if (args) {
                args += `&code=${azureFunctionsKey}`;
            }
            else {
                args = `?code=${azureFunctionsKey}`;
            }
        }
        return new Promise((resolve, reject) => {
            const requestData = {
                url: azureFunctionsUrl + funcName + '/' + args,
                /*          TODO - auth
                    headers: {
                        'Cookie' : this.credentials.Cookiestring(),
                    },*/
                /* TODO - params
                    body: {
                        name: name,
                        LuisAuthKey: luisKey
                    },
                    */
                json: true
            };
            CLDebug_1.CLDebug.LogRequest('GET', apiPath, requestData);
            Request.get(requestData, (error, response, body) => {
                if (error) {
                    reject(error);
                }
                else if (response.statusCode && response.statusCode >= 300) {
                    reject(body);
                }
                else {
                    resolve(body.Result);
                }
            });
        });
    }
}
exports.AzureFunctions = AzureFunctions;
