// @conversationlearner/models/dist/conversationlearner-models.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
	typeof define === 'function' && define.amd ? define(['exports'], factory) :
	(factory((global.conversationlearnerModels = {})));
}(this, (function (exports) { 'use strict';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = Object.setPrototypeOf ||
    ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
    function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };

function __extends(d, b) {
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var __assign = Object.assign || function __assign(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
    }
    return t;
};

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
}

/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
// TODO: Copied from ConversationLearner-UI. Need to improve Action design to find out how to consolidate knowledge to single layer
// Originally ConversationLearner-models was intended to know have to understand the tree structure which is why we added the
// pre-serialized 'text' field; however, now can't use the text field becuase it includes entity id's instead of human-readable names
// Instead of having to decide wether to perform substitutions from the serialized text or from the tree it's better to
// just have single method used everywhere
// Can't import because it's from ConversationLearner-UI
//import { NodeTypes } from "./models";
var NodeTypes;
(function (NodeTypes) {
    NodeTypes["Mention"] = "mention-inline-node";
    NodeTypes["Optional"] = "optional-inline-node";
})(NodeTypes || (NodeTypes = {}));
var defaultOptions = {
    fallbackToOriginal: false,
    preserveOptionalNodeWrappingCharacters: false
};
function serialize(value, entityValuesMap, userOptions) {
    if (userOptions === void 0) { userOptions = {}; }
    var options = __assign({}, defaultOptions, userOptions);
    var valueAsJson = typeof value.toJSON === 'function' ? value.toJSON() : value;
    var processedDocument = removeOptionalNodesWithoutEntityValues(valueAsJson.document, Array.from(entityValuesMap.keys()));
    return serializeNode(processedDocument, entityValuesMap, options);
}
/**
 * Given node return filter out optional nodes without matching values provided
 *
 * E.g. You are welcome[, $name] -> You are welcome
 * @param node Slate Node
 * @param entityValues Key Value pair of entity id to entity display value
 */
function removeOptionalNodesWithoutEntityValues(node, entityIds) {
    if (node.kind === 'inline' && node.type === NodeTypes.Optional) {
        var entityIdsWithinOptionalNode = getEntityIds(node);
        var hasValues = entityIdsWithinOptionalNode.every(function (x) { return entityIds.includes(x); });
        return hasValues ? node : undefined;
    }
    if (Array.isArray(node.nodes)) {
        node.nodes = node.nodes.map(function (n) { return removeOptionalNodesWithoutEntityValues(n, entityIds); }).filter(function (n) { return n; });
    }
    return node;
}
function getEntityIds(node) {
    var entityIds = [];
    // If current node is inline node which we know to have entityId then save it in the list
    if (node.kind === 'inline' && node.type === NodeTypes.Mention) {
        // This check is required because when input is Slate Value node is Immutable.Map object
        // but it could also be a node from value.toJSON()
        var data = typeof node.data.toJS === 'function' ? node.data.toJS() : node.data;
        var option = data.option;
        if (!option) {
            throw new Error("Attempting to serialize inline node but it did not have option");
        }
        var entityId = option.id;
        entityIds.push(entityId);
    }
    // Technically this would never get called because inline nodes shouldn't have other children which are inline nodes
    // however, it's good to have working depth-first-traversal anyways
    if (Array.isArray(node.nodes)) {
        var childrenEntityIds = node.nodes
            .map(function (n) { return getEntityIds(n); })
            .reduce(function (totalIds, nodeIds) { return totalIds.concat(nodeIds); }, []);
        entityIds.push.apply(entityIds, childrenEntityIds);
    }
    return entityIds;
}
function serializeNode(node, entityValues, options) {
    if (node.kind === 'text') {
        return node.leaves.map(function (n) { return n.text; }).join('');
    }
    var serializedChildNodes = node.nodes.map(function (n) { return serializeNode(n, entityValues, options); });
    if (node.kind === 'inline' && node.type === NodeTypes.Mention) {
        // This check is required because when input is Slate Value node is Immutable.Map object
        // but it could also be a node from value.toJSON()
        var data = typeof node.data.toJS === 'function' ? node.data.toJS() : node.data;
        if (!data.completed) {
            return serializedChildNodes.join('');
        }
        var option = data.option;
        if (!option) {
            throw new Error("Attempting to serialize inline node but it did not have option");
        }
        var entityId = option.id;
        var mapContainsEntity = entityValues.has(entityId);
        if (!mapContainsEntity) {
            if (options.fallbackToOriginal) {
                return serializedChildNodes.join('');
            }
            var entityValuesString = Array.from(entityValues.entries())
                .map(function (_a) {
                var id = _a[0], value = _a[1];
                return id + ": " + value;
            })
                .join(', ');
            throw new Error("Inline node representing entity " + entityId + " was NOT provided a value in the given entityValue map: [" + entityValuesString + "]");
        }
        return entityValues.get(entityId);
    }
    if (node.kind === 'document') {
        return serializedChildNodes.join('\n');
    }
    var serializedChildren = serializedChildNodes.join('');
    return node.kind === 'inline' && node.type === NodeTypes.Optional
        ? options.preserveOptionalNodeWrappingCharacters
            ? serializedChildren
            : serializedChildren.slice(1, -1)
        : serializedChildren;
}
var EntityIdSerializer = {
    serialize: serialize
};

/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
var ActionTypes = {
    TEXT: 'TEXT',
    API_LOCAL: 'API_LOCAL',
    // API_AZURE : "API_AZURE", TODO
    CARD: 'CARD'
};
var ActionBase = /** @class */ (function () {
    function ActionBase(init) {
        Object.assign(this, init);
    }
    // TODO: Refactor away from generic GetPayload for different action types
    // They all return strings but the strings are very different (Text is the substituted values, but other actions dont)
    // This causes issue of having to pass in entityValueMap even when it's not required, but making it optional ruins
    // safety for those places which should require it.
    // TODO: Remove ScoredAction since it doesn't have payload
    ActionBase.GetPayload = function (action, entityValues) {
        if (action.actionType === ActionTypes.TEXT) {
            /**
             * For backwards compatibility check if payload is of new TextPayload type
             * Ideally we would implement schema refactor:
             * 1. Make payloads discriminated unions (E.g. After checking the action.type, flow control knows the type of the payload property)
             * This removes the need for the GetPayload function and GetArguments which are brittle coding patterns.
             */
            try {
                var textPayload = JSON.parse(action.payload);
                return EntityIdSerializer.serialize(textPayload.json, entityValues);
            }
            catch (e) {
                var error = e;
                throw new Error("Error when attempting to parse text action payload. This might be an old action which was saved as a string.  Please create a new action. " + error.message);
            }
        }
        else if ([ActionTypes.CARD, ActionTypes.API_LOCAL].includes(action.actionType)) {
            var actionPayload = JSON.parse(action.payload);
            return actionPayload.payload;
        }
        return action.payload;
    };
    /** Return arguments for an action */
    ActionBase.GetActionArguments = function (action) {
        if (action.actionType !== ActionTypes.TEXT) {
            var actionPayload = JSON.parse(action.payload);
            return actionPayload.arguments.map(function (aa) { return new ActionArgument(aa); });
        }
        return [];
    };
    return ActionBase;
}());
var ActionArgument = /** @class */ (function () {
    function ActionArgument(actionArgument) {
        this.parameter = actionArgument.parameter;
        this.value = actionArgument.value.json;
    }
    ActionArgument.prototype.renderValue = function (entityValues, serializerOptions) {
        if (serializerOptions === void 0) { serializerOptions = {}; }
        return EntityIdSerializer.serialize(this.value, entityValues, serializerOptions);
    };
    return ActionArgument;
}());
var TextAction = /** @class */ (function (_super) {
    __extends(TextAction, _super);
    function TextAction(action) {
        var _this = _super.call(this, action) || this;
        if (action.actionType !== ActionTypes.TEXT) {
            throw new Error("You attempted to create text action from action of type: " + action.actionType);
        }
        _this.value = JSON.parse(_this.payload).json;
        return _this;
    }
    TextAction.prototype.renderValue = function (entityValues, serializerOptions) {
        if (serializerOptions === void 0) { serializerOptions = {}; }
        return EntityIdSerializer.serialize(this.value, entityValues, serializerOptions);
    };
    return TextAction;
}(ActionBase));
var ApiAction = /** @class */ (function (_super) {
    __extends(ApiAction, _super);
    function ApiAction(action) {
        var _this = _super.call(this, action) || this;
        if (action.actionType !== ActionTypes.API_LOCAL) {
            throw new Error("You attempted to create api action from action of type: " + action.actionType);
        }
        var actionPayload = JSON.parse(_this.payload);
        _this.name = actionPayload.payload;
        _this.arguments = actionPayload.arguments.map(function (aa) { return new ActionArgument(aa); });
        return _this;
    }
    ApiAction.prototype.renderArguments = function (entityValues, serializerOptions) {
        if (serializerOptions === void 0) { serializerOptions = {}; }
        return this.arguments.map(function (aa) {
            var value = null;
            try {
                value = EntityIdSerializer.serialize(aa.value, entityValues, serializerOptions);
            }
            catch (error) {
                // Just return null if argument doesn't have a value
            }
            return __assign({}, aa, { value: value });
        });
    };
    return ApiAction;
}(ActionBase));
var CardAction = /** @class */ (function (_super) {
    __extends(CardAction, _super);
    function CardAction(action) {
        var _this = _super.call(this, action) || this;
        if (action.actionType !== ActionTypes.CARD) {
            throw new Error("You attempted to create card action from action of type: " + action.actionType);
        }
        var actionPayload = JSON.parse(_this.payload);
        _this.templateName = actionPayload.payload;
        _this.arguments = actionPayload.arguments.map(function (aa) { return new ActionArgument(aa); });
        return _this;
    }
    CardAction.prototype.renderArguments = function (entityValues, serializerOptions) {
        if (serializerOptions === void 0) { serializerOptions = {}; }
        return this.arguments.map(function (aa) {
            var value = null;
            try {
                value = EntityIdSerializer.serialize(aa.value, entityValues, serializerOptions);
            }
            catch (error) {
                // Just return null if argument doesn't have a value
            }
            return __assign({}, aa, { value: value });
        });
    };
    return CardAction;
}(ActionBase));

(function (TrainingStatusCode) {
    TrainingStatusCode["Queued"] = "queued";
    TrainingStatusCode["Running"] = "running";
    TrainingStatusCode["Completed"] = "completed";
    TrainingStatusCode["Failed"] = "failed";
})(exports.TrainingStatusCode || (exports.TrainingStatusCode = {}));

/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

(function (EntityType) {
    EntityType["LOCAL"] = "LOCAL";
    EntityType["LUIS"] = "LUIS";
})(exports.EntityType || (exports.EntityType = {}));
var makeNegative = function (entity, positiveId) { return (__assign({}, entity, { negativeId: null, positiveId: positiveId })); };

var ModelUtils = /** @class */ (function () {
    function ModelUtils() {
    }
    /** Remove n words from start of string */
    ModelUtils.RemoveWords = function (text, numWords) {
        if (text.length === 0 || numWords === 0) {
            return text;
        }
        var firstSpace = text.indexOf(' ');
        var remaining = firstSpace > 0 ? text.slice(firstSpace + 1) : '';
        numWords--;
        return this.RemoveWords(remaining, numWords);
    };
    //====================================================================
    // CONVERSION: LabeledEntity == PredictedEntity
    //====================================================================
    ModelUtils.ToLabeledEntity = function (predictedEntity) {
        var score = predictedEntity.score, labeledEntity = __rest(predictedEntity, ["score"]);
        return predictedEntity;
    };
    ModelUtils.ToLabeledEntities = function (predictedEntities) {
        var labeledEntities = [];
        for (var _i = 0, predictedEntities_1 = predictedEntities; _i < predictedEntities_1.length; _i++) {
            var predictedEntity = predictedEntities_1[_i];
            var labelEntity = ModelUtils.ToLabeledEntity(predictedEntity);
            labeledEntities.push(labelEntity);
        }
        return labeledEntities;
    };
    ModelUtils.ToPredictedEntity = function (labeledEntity) {
        return __assign({}, labeledEntity, { score: undefined });
    };
    ModelUtils.ToPredictedEntities = function (labeledEntities) {
        var predictedEntities = [];
        for (var _i = 0, labeledEntities_1 = labeledEntities; _i < labeledEntities_1.length; _i++) {
            var labeledEntity = labeledEntities_1[_i];
            var predictedEntity = ModelUtils.ToPredictedEntity(labeledEntity);
            predictedEntities.push(predictedEntity);
        }
        return predictedEntities;
    };
    //====================================================================
    // CONVERSION: ExtractResponse == TextVariation
    //====================================================================
    ModelUtils.ToTextVariation = function (extractResponse) {
        var labeledEntities = this.ToLabeledEntities(extractResponse.predictedEntities);
        var textVariation = {
            text: extractResponse.text,
            labelEntities: labeledEntities
        };
        return textVariation;
    };
    ModelUtils.ToExtractResponses = function (textVariations) {
        var extractResponses = [];
        for (var _i = 0, textVariations_1 = textVariations; _i < textVariations_1.length; _i++) {
            var textVariation = textVariations_1[_i];
            var predictedEntities = this.ToPredictedEntities(textVariation.labelEntities);
            var extractResponse = {
                definitions: {
                    entities: [],
                    actions: [],
                    trainDialogs: []
                },
                packageId: '',
                metrics: {
                    wallTime: 0
                },
                text: textVariation.text,
                predictedEntities: predictedEntities
            };
            extractResponses.push(extractResponse);
        }
        return extractResponses;
    };
    ModelUtils.ToTextVariations = function (extractResponses) {
        var textVariations = [];
        for (var _i = 0, extractResponses_1 = extractResponses; _i < extractResponses_1.length; _i++) {
            var extractResponse = extractResponses_1[_i];
            var labelEntities = this.ToLabeledEntities(extractResponse.predictedEntities);
            var textVariation = {
                text: extractResponse.text,
                labelEntities: labelEntities
            };
            textVariations.push(textVariation);
        }
        return textVariations;
    };
    //====================================================================
    // CONVERSION: LogDialog == TrainDialog
    //====================================================================
    ModelUtils.ToTrainDialog = function (logDialog, actions, entities) {
        if (actions === void 0) { actions = null; }
        if (entities === void 0) { entities = null; }
        var trainRounds = [];
        for (var _i = 0, _a = logDialog.rounds; _i < _a.length; _i++) {
            var logRound = _a[_i];
            var trainRound = ModelUtils.ToTrainRound(logRound);
            trainRounds.push(trainRound);
        }
        var appDefinition = null;
        if (entities != null && actions != null) {
            appDefinition = {
                entities: entities,
                actions: actions,
                trainDialogs: []
            };
        }
        return {
            packageCreationId: 0,
            packageDeletionId: 0,
            trainDialogId: '',
            sourceLogDialogId: logDialog.logDialogId,
            version: 0,
            rounds: trainRounds,
            definitions: appDefinition
        };
    };
    //====================================================================
    // CONVERSION: LogRoung == TrainRound
    //====================================================================
    ModelUtils.ToTrainRound = function (logRound) {
        return {
            extractorStep: {
                textVariations: [
                    {
                        labelEntities: ModelUtils.ToLabeledEntities(logRound.extractorStep.predictedEntities),
                        text: logRound.extractorStep.text
                    }
                ]
            },
            scorerSteps: logRound.scorerSteps.map(function (logScorerStep) { return ({
                input: logScorerStep.input,
                labelAction: logScorerStep.predictedAction,
                scoredAction: undefined
            }); })
        };
    };
    //====================================================================
    // CONVERSION: LogScorerStep == TrainScorerStep
    //====================================================================
    ModelUtils.ToTrainScorerStep = function (logScorerStep) {
        return {
            input: logScorerStep.input,
            labelAction: logScorerStep.predictedAction,
            scoredAction: undefined
        };
    };
    //====================================================================
    // CONVERSION: TrainDialog == CreateTeachParams
    //====================================================================
    ModelUtils.ToCreateTeachParams = function (trainDialog) {
        var createTeachParams = {
            contextDialog: trainDialog.rounds,
            sourceLogDialogId: trainDialog.sourceLogDialogId
        };
        // TODO: Change to non destructive operation
        // Strip out "entityType" (*sigh*)
        for (var _i = 0, _a = createTeachParams.contextDialog; _i < _a.length; _i++) {
            var round = _a[_i];
            for (var _b = 0, _c = round.extractorStep.textVariations; _b < _c.length; _b++) {
                var textVariation = _c[_b];
                for (var _d = 0, _e = textVariation.labelEntities; _d < _e.length; _d++) {
                    var labeledEntity = _e[_d];
                    delete labeledEntity.entityType;
                }
            }
        }
        return createTeachParams;
    };
    //====================================================================
    // CONVERSION: TeachResponse == Teach
    //====================================================================
    ModelUtils.ToTeach = function (teachResponse) {
        return {
            teachId: teachResponse.teachId,
            trainDialogId: teachResponse.trainDialogId,
            createdDatetime: undefined,
            lastQueryDatetime: undefined,
            packageId: undefined
        };
    };
    ModelUtils.PrebuiltDisplayText = function (prebuiltType, resolution, entityText) {
        if (prebuiltType.startsWith('builtin.encyclopedia')) {
            return entityText;
        }
        switch (prebuiltType) {
            case 'builtin.datetimeV2.date':
                var date = resolution.values[0].value;
                if (resolution.values[1]) {
                    date += " or " + resolution.values[1].value;
                }
                return date;
            case 'builtin.datetimeV2.time':
                var time = resolution.values[0].value;
                if (resolution.values[1]) {
                    time += " or " + resolution.values[1].value;
                }
                return time;
            case 'builtin.datetimeV2.daterange':
                return resolution.values[0].start + " to " + resolution.values[0].end;
            case 'builtin.datetimeV2.timerange':
                return resolution.values[0].start + " to " + resolution.values[0].end;
            case 'builtin.datetimeV2.datetimerange':
                return resolution.values[0].start + " to " + resolution.values[0].end;
            case 'builtin.datetimeV2.duration':
                return resolution.values[0].value + " seconds";
            case 'builtin.datetimeV2.set':
                return "" + resolution.values[0].value;
            case 'builtin.number':
                return resolution.value;
            case 'builtin.ordinal':
                return resolution.value;
            case 'builtin.temperature':
                return resolution.value;
            case 'builtin.dimension':
                return resolution.value;
            case 'builtin.money':
                return resolution.value;
            case 'builtin.age':
                return resolution.value;
            case 'builtin.percentage':
                return resolution.value;
            case 'builtin.geography.city':
                return resolution.value;
            case 'builtin.geography.country':
                return resolution.value;
            case 'builtin.geography.pointOfInterest':
                return resolution.value;
            case 'builtin.encyclopedia':
                return entityText;
            default:
                return entityText;
        }
    };
    return ModelUtils;
}());

var filledEntityValueAsString = function (fe) { return memoryValuesAsString(fe.values); };
var memoryValuesAsString = function (memories) {
    // Print out list in friendly manner
    var group = '';
    for (var key in memories) {
        var index = +key;
        var prefix = '';
        if (memories.length !== 1 && index === memories.length - 1) {
            prefix = ' and ';
        }
        else if (index !== 0) {
            prefix = ', ';
        }
        var value = memories[key];
        var text = value.displayText ? value.displayText : value.userText;
        group += "" + prefix + text;
    }
    return group;
};
// In future will refactor to remove FilledEntityMap, but this method provides abstraction
// or isolation to methods that would normally consume filledEntityMap to reduce coupling
// and make code more flexible
var getEntityDisplayValueMap = function (filledEntityMap) {
    return Object.keys(filledEntityMap.map).reduce(function (m, key) {
        var entityDisplayValue = filledEntityMap.ValueAsString(key);
        // TODO: Required check because poor API from filledEntityMap which can return null
        if (entityDisplayValue) {
            m.set(key, entityDisplayValue);
        }
        return m;
    }, new Map());
};
var FilledEntityMap = /** @class */ (function () {
    function FilledEntityMap(init) {
        this.map = {};
        Object.assign(this, init);
    }
    FilledEntityMap.prototype.ValueAsList = function (entityName) {
        if (!this.map[entityName]) {
            return [];
        }
        return this.map[entityName].values.filter(function (v) { return typeof v.userText === 'string'; }).map(function (v) { return v.userText; });
    };
    FilledEntityMap.prototype.ValueAsString = function (entityName) {
        if (!this.map[entityName]) {
            return null;
        }
        // Print out list in friendly manner
        return filledEntityValueAsString(this.map[entityName]);
    };
    FilledEntityMap.prototype.ValueAsNumber = function (entityName) {
        var textObj = this.ValueAsString(entityName);
        var number = Number(textObj);
        if (isNaN(number)) {
            return null;
        }
        return number;
    };
    FilledEntityMap.prototype.ValueAsBoolean = function (entityName) {
        var textObj = this.ValueAsString(entityName);
        if (textObj) {
            if (textObj.toLowerCase() === 'true') {
                return true;
            }
            if (textObj.toLowerCase() === 'false') {
                return false;
            }
        }
        return null;
    };
    FilledEntityMap.prototype.ValueAsObject = function (entityName) {
        var textObj = this.ValueAsString(entityName);
        if (textObj) {
            return JSON.parse(textObj);
        }
        return null;
    };
    FilledEntityMap.prototype.ValueAsPrebuilt = function (entityName) {
        if (!this.map[entityName]) {
            return [];
        }
        return this.map[entityName].values;
    };
    FilledEntityMap.prototype.Forget = function (entityName, entityValue, isBucket) {
        if (entityValue === void 0) { entityValue = null; }
        if (isBucket === void 0) { isBucket = false; }
        // Check if entity buckets values
        if (isBucket) {
            // Entity might not be in memory
            if (!this.map[entityName]) {
                return;
            }
            // If no entity Value provide, clear the entity
            if (!entityValue) {
                delete this.map[entityName];
            }
            else {
                // Find case insensitive index
                var lowerCaseNames = this.map[entityName].values
                    .filter(function (mv) { return mv.userText; })
                    .map(function (mv) { return mv.userText.toLowerCase(); });
                var index = lowerCaseNames.indexOf(entityValue.toLowerCase());
                if (index > -1) {
                    this.map[entityName].values.splice(index, 1);
                    if (this.map[entityName].values.length === 0) {
                        delete this.map[entityName];
                    }
                }
            }
        }
        else {
            delete this.map[entityName];
        }
    };
    // Remember multiple values for an entity
    FilledEntityMap.prototype.RememberMany = function (entityName, entityId, entityValues, isBucket, builtinType, resolution) {
        if (isBucket === void 0) { isBucket = false; }
        if (builtinType === void 0) { builtinType = null; }
        if (resolution === void 0) { resolution = null; }
        for (var _i = 0, entityValues_1 = entityValues; _i < entityValues_1.length; _i++) {
            var entityValue = entityValues_1[_i];
            this.Remember(entityName, entityId, entityValue, isBucket, builtinType, resolution);
        }
    };
    // Remember value for an entity 
    FilledEntityMap.prototype.Remember = function (entityName, entityId, entityValue, isBucket, builtinType, resolution) {
        if (isBucket === void 0) { isBucket = false; }
        if (builtinType === void 0) { builtinType = null; }
        if (resolution === void 0) { resolution = null; }
        if (!this.map[entityName]) {
            this.map[entityName] = {
                entityId: entityId,
                values: []
            };
        }
        var displayText = builtinType ? ModelUtils.PrebuiltDisplayText(builtinType, resolution, entityValue) : entityValue;
        var filledEntity = this.map[entityName];
        // Check if entity buckets values
        if (isBucket) {
            // Add if not a duplicate
            var containsDuplicateValue = filledEntity.values.some(function (memoryValue) { return memoryValue.userText === entityValue; });
            if (!containsDuplicateValue) {
                filledEntity.values.push({
                    userText: entityValue,
                    displayText: displayText,
                    builtinType: builtinType,
                    resolution: resolution
                });
            }
        }
        else {
            filledEntity.values = [{ userText: entityValue, displayText: displayText, builtinType: builtinType, resolution: resolution }];
        }
    };
    /** Return FilledEntity array for items I've remembered */
    FilledEntityMap.prototype.FilledEntities = function () {
        var _this = this;
        return Object.keys(this.map).map(function (val) {
            return _this.map[val];
        });
    };
    FilledEntityMap.Split = function (action) {
        return action.split(/[\s,:.?!\[\]]+/);
    };
    return FilledEntityMap;
}());

/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
var KeyGen = /** @class */ (function () {
    function KeyGen() {
    }
    KeyGen.MakeKey = function (address) {
        return this.HashCode(address).toString();
    };
    KeyGen.HashCode = function (text) {
        var hash = 0;
        var i = undefined;
        var chr = undefined;
        if (text.length === 0)
            return hash;
        for (i = 0; i < text.length; i++) {
            chr = text.charCodeAt(i);
            hash = (hash << 5) - hash + chr;
            hash |= 0; // Convert to 32bit integer
        }
        return hash;
    };
    return KeyGen;
}());

/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

(function (ReplayErrorType) {
    ReplayErrorType["MissingAction"] = "MissingAcion";
    ReplayErrorType["MissingEntity"] = "MissingEntity";
    ReplayErrorType["ActionUnavailable"] = "ActionUnavailable";
    ReplayErrorType["EntityDiscrepancy"] = "EntityDiscrepancy";
})(exports.ReplayErrorType || (exports.ReplayErrorType = {}));
var ReplayError = /** @class */ (function () {
    function ReplayError(type) {
        this.type = type;
    }
    return ReplayError;
}());
var ReplayErrorMissingAction = /** @class */ (function (_super) {
    __extends(ReplayErrorMissingAction, _super);
    function ReplayErrorMissingAction(lastUserInput) {
        var _this = _super.call(this, exports.ReplayErrorType.MissingAction) || this;
        _this.lastUserInput = lastUserInput;
        return _this;
    }
    return ReplayErrorMissingAction;
}(ReplayError));
var ReplayErrorMissingEntity = /** @class */ (function (_super) {
    __extends(ReplayErrorMissingEntity, _super);
    function ReplayErrorMissingEntity(value) {
        var _this = _super.call(this, exports.ReplayErrorType.MissingEntity) || this;
        _this.value = value;
        return _this;
    }
    return ReplayErrorMissingEntity;
}(ReplayError));
var ReplayErrorActionUnavailable = /** @class */ (function (_super) {
    __extends(ReplayErrorActionUnavailable, _super);
    function ReplayErrorActionUnavailable(lastUserInput) {
        var _this = _super.call(this, exports.ReplayErrorType.ActionUnavailable) || this;
        _this.lastUserInput = lastUserInput;
        return _this;
    }
    return ReplayErrorActionUnavailable;
}(ReplayError));
var ReplayErrorEntityDiscrepancy = /** @class */ (function (_super) {
    __extends(ReplayErrorEntityDiscrepancy, _super);
    function ReplayErrorEntityDiscrepancy(lastUserInput, originalEntities, newEntities) {
        var _this = _super.call(this, exports.ReplayErrorType.EntityDiscrepancy) || this;
        _this.lastUserInput = lastUserInput;
        _this.originalEntities = originalEntities;
        _this.newEntities = newEntities;
        return _this;
    }
    return ReplayErrorEntityDiscrepancy;
}(ReplayError));

(function (DialogType) {
    DialogType["TEACH"] = "teach";
    DialogType["TRAINDIALOG"] = "traindialog";
    DialogType["LOGDIALOG"] = "logdialog";
})(exports.DialogType || (exports.DialogType = {}));

(function (DialogMode) {
    DialogMode["Extractor"] = "Extract";
    DialogMode["Scorer"] = "Score";
    DialogMode["Wait"] = "Wait"; // Waiting for user input
})(exports.DialogMode || (exports.DialogMode = {}));

(function (SenderType) {
    SenderType[SenderType["User"] = 0] = "User";
    SenderType[SenderType["Bot"] = 1] = "Bot";
})(exports.SenderType || (exports.SenderType = {}));

var CL_USER_NAME_ID = 'CLTrainer';
var MEMORY_KEY_HEADER_NAME = 'x-conversationlearner-memory-key';

(function (ScoreReason) {
    // Action has been masked
    ScoreReason["NotAvailable"] = "notAvailable";
    // Action can't be scored because it hasn't been trained yet
    ScoreReason["NotScorable"] = "notScorable";
    // Score has not yet been calculated
    ScoreReason["NotCalculated"] = "notCalculated";
})(exports.ScoreReason || (exports.ScoreReason = {}));

/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

exports.EntityIdSerializer = EntityIdSerializer;
exports.ActionTypes = ActionTypes;
exports.ActionBase = ActionBase;
exports.ActionArgument = ActionArgument;
exports.TextAction = TextAction;
exports.ApiAction = ApiAction;
exports.CardAction = CardAction;
exports.makeNegative = makeNegative;
exports.filledEntityValueAsString = filledEntityValueAsString;
exports.memoryValuesAsString = memoryValuesAsString;
exports.getEntityDisplayValueMap = getEntityDisplayValueMap;
exports.FilledEntityMap = FilledEntityMap;
exports.KeyGen = KeyGen;
exports.ModelUtils = ModelUtils;
exports.ReplayError = ReplayError;
exports.ReplayErrorMissingAction = ReplayErrorMissingAction;
exports.ReplayErrorMissingEntity = ReplayErrorMissingEntity;
exports.ReplayErrorActionUnavailable = ReplayErrorActionUnavailable;
exports.ReplayErrorEntityDiscrepancy = ReplayErrorEntityDiscrepancy;
exports.CL_USER_NAME_ID = CL_USER_NAME_ID;
exports.MEMORY_KEY_HEADER_NAME = MEMORY_KEY_HEADER_NAME;

Object.defineProperty(exports, '__esModule', { value: true });

})));
